File: leetcode/101-200/107-binary-tree-level-order-traversal-ii/bench/level_order_bottom.py
#!/usr/bin/env python3
# Benchmark workload for LeetCode #107 — bottom-up level order, Python mirror.
# Runs a SMALLER K (pure-Python recursion is slow); timed separately, NOT cross-checked.
import sys

MOD = 1000000007


class Node:
    __slots__ = ("val", "left", "right")

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None


def insert(root, v):
    if root is None:
        return Node(v)
    if v < root.val:
        root.left = insert(root.left, v)
    else:
        root.right = insert(root.right, v)
    return root


def dfs(node, depth, rows):
    if node is None:
        return
    if depth == len(rows):
        rows.append([])
    rows[depth].append(node.val)
    dfs(node.left, depth + 1, rows)
    dfs(node.right, depth + 1, rows)


def level_order_bottom(root):
    rows = []
    dfs(root, 0, rows)
    out = []
    for d in range(len(rows) - 1, -1, -1):
        out.append(list(rows[d]))
    return out


def main():
    base = [16, 8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30]
    bn = len(base)
    pool = []
    for t in range(8):
        root = None
        for k in range(bn):
            root = insert(root, base[(k + t) % bn])
        pool.append(root)
    acc = 1
    K = 50000
    for _ in range(K):
        idx = acc % 8
        levels = level_order_bottom(pool[idx])
        acc = (acc * 131 + len(levels)) % MOD
        for lvl in levels:
            acc = (acc * 131 + len(lvl)) % MOD
            for v in lvl:
                acc = (acc * 131 + v) % MOD
    print(acc)


if __name__ == "__main__":
    sys.setrecursionlimit(10000)
    main()

File: leetcode/101-200/107-binary-tree-level-order-traversal-ii/bench/level_order_bottom.rs
// Benchmark workload for LeetCode #107 — bottom-up level order, Rust mirror (Box + &-borrow).
// The tree is never mutated or shared during traversal, so a single-owner Box walked by
// reference is the honest Rust node. Build 8 BSTs once, then K reps of DFS-with-depth bottom-up
// level order on a data-dependent-selected tree, building a fresh Vec<Vec<i64>> output (rows
// deepest-first, mirroring the kara `out` allocation) and folding it into a rolling hash.
const MOD: i64 = 1000000007;

struct Node { val: i64, left: Option<Box<Node>>, right: Option<Box<Node>> }

fn insert(root: Option<Box<Node>>, v: i64) -> Option<Box<Node>> {
    match root {
        None => Some(Box::new(Node { val: v, left: None, right: None })),
        Some(mut n) => {
            if v < n.val { n.left = insert(n.left.take(), v); }
            else { n.right = insert(n.right.take(), v); }
            Some(n)
        }
    }
}
fn dfs(node: &Option<Box<Node>>, depth: usize, rows: &mut Vec<Vec<i64>>) {
    if let Some(n) = node {
        if depth == rows.len() { rows.push(Vec::new()); }
        rows[depth].push(n.val);
        dfs(&n.left, depth + 1, rows);
        dfs(&n.right, depth + 1, rows);
    }
}
fn level_order_bottom(root: &Option<Box<Node>>) -> Vec<Vec<i64>> {
    let mut rows: Vec<Vec<i64>> = Vec::new();
    dfs(root, 0, &mut rows);
    let mut out: Vec<Vec<i64>> = Vec::new();
    for d in (0..rows.len()).rev() {
        let mut newrow: Vec<i64> = Vec::new();
        for &v in &rows[d] { newrow.push(v); }
        out.push(newrow);
    }
    out
}

fn main() {
    let base: [i64; 15] = [16, 8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30];
    let bn = base.len();
    let mut pool: Vec<Option<Box<Node>>> = Vec::new();
    for t in 0..8 {
        let mut root: Option<Box<Node>> = None;
        for k in 0..bn { root = insert(root, base[(k + t) % bn]); }
        pool.push(root);
    }
    let mut acc: i64 = 1;
    for _ in 0..1000000i64 {
        let idx = (acc % 8) as usize;
        let levels = level_order_bottom(&pool[idx]);
        acc = (acc * 131 + levels.len() as i64) % MOD;
        for lvl in &levels {
            acc = (acc * 131 + lvl.len() as i64) % MOD;
            for &v in lvl {
                acc = (acc * 131 + v) % MOD;
            }
        }
    }
    println!("{}", acc);
}
